### PR TITLE
HDDS-9076. Recon UI OMDB Insights Changes and Improvements

### PR DESCRIPTION
**Pull request by smitajoshi12**
## What changes were proposed in this pull request?

1) Take user to Key Deletion page on click of "Overview" pending deletion card.
2) Take user to Pending Open Keys page on click of "Overview" pending open keys.
3) API param "prevKey" of deletePending keys and open keys seems not passing value.
4) First Time in API Call no need to call prevKey as we sending empty.
5) Pending Key Deletion from Overview Page need to delete.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-9076

## How was this patch tested?
Manually

[1.webm](https://github.com/apache/ozone/assets/112169209/45cc847a-3168-451f-9d16-c19c9fc6eeaf)

image

image

### CODE REVIEW

**Comment by dombizita** [comment]
could you take a look @devmadhuu, @ArafatKhan2198 and @devabhishekpal? thanks in advance!

**Review by dombizita — COMMENTED** [review]
hi @smitajoshi12, thanks for working on this! could you please check if there are any necessary changes after #5023? if yes, maybe it should be separated and solved in another jira, once that's done you can continue on this.

**Comment by smitajoshi12** [comment]
> hi @smitajoshi12, thanks for working on this! could you please check if there are any necessary changes after #5023? if yes, maybe it should be separated and solved in another jira, once that's done you can continue on this.

@dombizita  Yes I will check with @ArafatKhan2198

**Comment by ArafatKhan2198** [comment]
> hi @smitajoshi12, thanks for working on this! could you please check if there are any necessary changes after #5023? if yes, maybe it should be separated and solved in another jira, once that's done you can continue on this.

I ensured that the frontend was changed to align with the modifications made to the backend API endpoints. Therefore, I am confident that there shouldn't be any issues 👍

**Review by devmadhuu — APPROVED** [review]
LGTM +1

**Comment by ArafatKhan2198** [comment]
@devabhishekpal Could you take a look at it once

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx [hunk 1]
```diff
@@ -46,18 +46,31 @@ const defaultProps = {
...
   linkToUrl: string;
+  title:string
 }
 
 class OverviewCardWrapper extends React.Component<IOverviewCardWrapperProps> {
   render() {
     const {linkToUrl, children} = this.props;
-    if (linkToUrl) {
+    if (linkToUrl ) {
+      if(linkToUrl === '/Om'){
```
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx:56]: Can we shorten this check? Instead of spanning two lines can you try to make the check as:
```
if (linkToUrl){
    return (
        <Link to={{
            pathname: linkToUrl
            state: { linkToUrl === "/Om" ? { activeTab: .... } : undefined }
            }}
        />
    )
}
```

Could you check this out?

Comment (devabhishekpal) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx:56]: Rest looks good to me

Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx:56]: @devabhishekpal 
It should not allow to enter and use  <Link > tag so gave additional condition to mimimize computation time.

Comment (devabhishekpal) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx:56]: Okay, looks good to me

**Comment by ArafatKhan2198** [comment]
Thanks for working on this @smitajoshi12 LGTM +1

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/om/om.tsx [hunk 2]
```diff
@@ -444,15 +443,25 @@ export class Om extends React.Component<Record<string, object>, IOmdbInsightsSta
...
+      this.fetchOpenKeys(this.state.includeFso, this.state.includeNonFso, this.state.DEFAULT_LIMIT, this.state.prevKeyOpen);
+    } else if (this.state.activeTab  === '3') {
+      keysPendingExpanded =[];
+      this.fetchDeletePendingKeys(this.state.DEFAULT_LIMIT, this.state.prevKeyDeletePending);
+    } else if (this.state.activeTab  === '4') {
+      this.fetchDeletedKeys(this.state.DEFAULT_LIMIT, this.state.prevKeyDeleted);
+    }
+    else {
+      this.fetchMismatchContainers(this.state.DEFAULT_LIMIT, this.state.prevKeyMismatch, this.state.mismatchMissingState);
+    }
```
[the remark this answers is not in the document]
Comment (dombizita) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/om/om.tsx:456]: or is it possible, that the `this.state.activeTab` is not set here? can it be different from 1, 2, 3 or 4 (if we have 4 tabs)?

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/om/om.tsx [hunk 3]
```diff
@@ -493,13 +502,15 @@ export class Om extends React.Component<Record<string, object>, IOmdbInsightsSta
   fetchOpenKeys = (includeFso: boolean, includeNonFso: boolean, limit: number, prevKeyOpen: string) => {
     this.setState({
       loading: true,
-      clickable: true,
-      prevClickable:true
+      nextClickable: true,
+      prevClickable: true,
+      includeFso,
+      includeNonFso
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/om/om.tsx:508]: can you help me understand why is this needed? I don't see any new changes related to this, was something not working correctly before?

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx [hunk 4]
```diff
@@ -46,18 +46,31 @@ const defaultProps = {
 
 interface IOverviewCardWrapperProps {
   linkToUrl: string;
+  title:string
```
[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx]: @dombizita 
Completed change in latest commit.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx]: @dombizita 
Completed Change in Latest Commit.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/om/om.tsx [hunk 5]
```diff
@@ -326,12 +326,12 @@ export class Om extends React.Component<Record<string, object>, IOmdbInsightsSta
       prevKeyDeletePending: "",
       prevKeyDeleted: 0,
       expandedRowData: {},
-      activeTab: '1',
+      activeTab: props.location.state && props.location.state.activeTab !== '1' ? props.location.state.activeTab:'1',
```
[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/om/om.tsx]: @dombizita 
Complteed change with props.location.state ? props.location.state.activeTab : '1',

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/om/om.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/om/om.tsx]: @dombizita 
I have added else if (this.state.activeTab  ===  '1') { so if we have to add additional tab then we can add with new number.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx [hunk 6]
```diff
@@ -46,18 +46,31 @@ const defaultProps = {
...
   render() {
     const {linkToUrl, children} = this.props;
-    if (linkToUrl) {
+    if (linkToUrl ) {
+      if(linkToUrl === '/Om'){
       return (
-        <Link to={linkToUrl}>
+        <Link to={{
+          pathname: linkToUrl,
+          state: { activeTab: children._owner && children._owner.stateNode.props && children._owner.stateNode.props.title === "Open Keys Summary"? '2':'3'}
```
[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx:60]: @dombizita 

Yes you are correct zita. If we add another card with a link to /Om to the Overview page, we can pass currentTab value to another component.

We will use key value pair for below condition if we have to add another tab will modify this condition accordingly. Will do next PR.

state: { activeTab: children._owner && children._owner.stateNode.props && children._owner.stateNode.props.title === "Open Keys Summary"? '2':'3'}

[the remark this answers is not in the document]
Comment (dombizita) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx:60]: can't we do something like this now?
```
if (children._owner && children._owner.stateNode.props) {
  let title = children._owner.stateNode.props.title;
  let activeTab = '1';
  if (title === "Open Keys Summary") {
    activeTab = '2'
  } else if (title === "Pending Deleted Keys Summary") {
    activeTab = '3'
  }
}
```
and later we just need this:
```
state: { activeTab: activeTab }
```
let me know what you think of this. with this approach if someone reads the code they would automatically know hot the active tab is set, without checking the code elsewhere. important: I didn't test this code, just a suggestion!

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/om/om.tsx [hunk 7]
```diff
@@ -444,15 +443,25 @@ export class Om extends React.Component<Record<string, object>, IOmdbInsightsSta
...
+      this.fetchOpenKeys(this.state.includeFso, this.state.includeNonFso, this.state.DEFAULT_LIMIT, this.state.prevKeyOpen);
+    } else if (this.state.activeTab  === '3') {
+      keysPendingExpanded =[];
+      this.fetchDeletePendingKeys(this.state.DEFAULT_LIMIT, this.state.prevKeyDeletePending);
+    } else if (this.state.activeTab  === '4') {
+      this.fetchDeletedKeys(this.state.DEFAULT_LIMIT, this.state.prevKeyDeleted);
+    }
+    else if (this.state.activeTab  === '1') {
+      this.fetchMismatchContainers(this.state.DEFAULT_LIMIT, this.state.prevKeyMismatch, this.state.mismatchMissingState);
+    }
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/om/om.tsx:456]: nit
```suggestion
    if (this.state.activeTab === '1') {
      this.fetchMismatchContainers(this.state.DEFAULT_LIMIT, this.state.prevKeyMismatch, this.state.mismatchMissingState);
    } else if (this.state.activeTab === '2') {
      this.fetchOpenKeys(this.state.includeFso, this.state.includeNonFso, this.state.DEFAULT_LIMIT, this.state.prevKeyOpen);
    } else if (this.state.activeTab === '3') {
      keysPendingExpanded =[];
      this.fetchDeletePendingKeys(this.state.DEFAULT_LIMIT, this.state.prevKeyDeletePending);
    } else if (this.state.activeTab === '4') {
      this.fetchDeletedKeys(this.state.DEFAULT_LIMIT, this.state.prevKeyDeleted);
    }
```
by the way why is `keysPendingExpanded =[];` needed here?

Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/om/om.tsx]: @dombizita 
Completed Code Sequencing. 
keysPendingExpanded =[]; is array of expanding element making empty whenever API call is getting called.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx]: @dombizita 
updated in Recent Commit.
 setCurrentActiveTab = (title: any) => {
    if (title === "Open Keys Summary") {
      return {
        active: '2'
      }
    }
    else if (title === "Pending Deleted Keys Summary") {
      return {
        active: '3'
      }
    }
  };
We can add new tab with new condition so made it generic.

**Comment by dombizita** [comment]
thanks for addressing the review comments @smitajoshi12! I tried your latest patch with `pnpm run dev` and the Overview page was not loading for me, this was the error message on the UI: 
```
Error while fetching data
AxiosError: Request failed with status code 404
```
here is a screenshot:

overview_page

could you please check it? 

also on the OM DB Insights I saw this, this is because the `db.json` has some keys duplicated?

om_db_insights

**Comment by ArafatKhan2198** [comment]
@smitajoshi12 Could you pull in the latest changes from upstream and see if everything is working out fine here?

> thanks for addressing the review comments @smitajoshi12! I tried your latest patch with `pnpm run dev` and the Overview page was not loading for me, this was the error message on the UI:
> 
> ```
> Error while fetching data
> AxiosError: Request failed with status code 404
> ```
> 
> here is a screenshot:
> 
> overview_page
> could you please check it?
> 
> also on the OM DB Insights I saw this, this is because the `db.json` has some keys duplicated?
> 
> om_db_insights

@dombizita I believe the error which was not thrown earlier is happening due to the changes brought in by this [PR](https://github.com/apache/ozone/pull/5154/files#diff-9446deb0dd230ce73af58b6a6efa055d491fba0816f973063923079e56e63efd)
The fix for this raised by @smitajoshi12 over here :- [Fix](https://github.com/apache/ozone/pull/5237)

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx [hunk 8]
```diff
@@ -46,18 +46,45 @@ const defaultProps = {
...
+        )
+      }
+      else{
+        return (
+            <Link to={linkToUrl}>
+              {children}
+            </Link>
+        )
+      }
+      }
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx:87]: could you please fix the indentation here? in this file it seems to be 2 spaces.

**Comment by ArafatKhan2198** [comment]
Thank you, @smitajoshi12, for rebasing to the latest changes! Once you've integrated Zita's latest comment, I believe we'll be ready to proceed.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx (hunk 8 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx]: @dombizita Done Default formatting and pushed code in latest commit.

**Review by ArafatKhan2198 — APPROVED** [review]

**Review by dombizita — APPROVED** [review]

**Comment by dombizita** [comment]
thanks for working on this @smitajoshi12! thanks for the review @ArafatKhan2198 and @devabhishekpal!
